sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -373,11 +378,22 @@ public WriterImpl(CassandraIO.Write<T> spec) {
     @Override
     public void write(T entity) {
       Mapper<T> mapper = (Mapper<T>) mappingManager.mapper(entity.getClass());
-      mapper.save(entity);
+      this.writeFutures.add(mapper.saveAsync(entity));
+      if (this.writeFutures.size() % CONCURRENT_ASYNC_QUERIES == 0) {
+        // We reached the max number of allowed in flight queries
+        LOG.debug("Waiting for a batch of 100 Cassandra writes to be executed...");
+        Futures.successfulAsList(this.writeFutures);
+        this.writeFutures = Lists.newArrayList();
+      }
     }
 
     @Override
     public void close() {
+      if (this.writeFutures.size() > 0) {
+        // Waiting for the last in flight queries to end
+        Futures.successfulAsList(this.writeFutures);
```

Comment:
The same question about errors handling - do we have to check that all entities were stored to Casandra successfully?

Reply:
Same answer, a loop over the futures will be more appropriate, nice catch.